kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -149,23 +151,22 @@ private void tryEnsureNextActionsIterIsReady() {
     }
 
     /**
-     * Get the next file from `filesIter` (.json or .checkpoint.parquet), contextualize it
+     * Get the next file from `filesList` (.json or .checkpoint.parquet), contextualize it
      * (allow the connector to split it), and then read it + inject the `isFromCheckpoint`
      * information.
      * <p>
-     * Requires that `filesIter.hasNext` is true.
+     * Requires that `filesList.isEmpty` is false.
      */
     private CloseableIterator<ActionWrapper> getNextActionsIter() {
-        final FileStatus nextFile = filesIter.next();
-
-        // TODO: [#1965] It should be possible to read our JSON and parquet files
-        //       many-at-once instead of one at a time.
-
+        final FileStatus nextFile = filesList.pop();
         try {
             if (nextFile.getPath().endsWith(".json")) {
                 final long fileVersion = FileNames.deltaVersion(nextFile.getPath());
 
-                // Read that file
+                // We can not read multiple JSON files in parallel (like the checkpoint files),
+                // because each one has a different version, and we need to associate the version
+                // with actions read from the JSON file for further optimizations later on.
+
                 final CloseableIterator<ColumnarBatch> dataIter =
                     tableClient.getJsonHandler().readJsonFiles(
```

Comment:
is this an optimization that we should consider? that the returned columnar batches has the file status of the file from which it was read?